Add NSwag.MSBuild.CodeGeneration package

dougbu opened the pull request:
- #1587, #1588

RSuter commented on src/NSwag.MSBuild.CodeGeneration/NSwag.MSBuild.CodeGeneration.targets:
```diff
@@ -0,0 +1,94 @@
+    <ItemGroup>
+      <!-- @(CurrentServiceProjectReference) item group will never contain more than one item. -->
+      <CurrentServiceProjectReference Update="@(CurrentServiceProjectReference)">
+        <Command>$(NSwagExe)</Command>
```
Shouldnt this also have a condition? Or is it overwritten by the next command?

dougbu replied:
Later settings of the `%(Command)` metadata override earlier ones. This is just he default.

dougbu commented on src/NSwag.MSBuild.CodeGeneration/NSwag.MSBuild.CodeGeneration.targets:
```diff
@@ -0,0 +1,94 @@
+        <Command Condition="'%(TargetFramework)' == 'netcoreapp1.0'">$(NSwagExe_Core10)</Command>
+        <Command Condition="'%(TargetFramework)' == 'netcoreapp1.1'">$(NSwagExe_Core11)</Command>
+        <Command Condition="'%(TargetFramework)' == 'netcoreapp2.0'">$(NSwagExe_Core20)</Command>
+        <Command Condition="'%(TargetFramework)' == 'netcoreapp2.1'">$(NSwagExe_Core21)</Command>
```
Since `netcoreapp2.2` is available, I should add a case for that too (using `$(NSwagExe_Core21)` for now).

RSuter replied:
Added new issue: https://github.com/RSuter/NSwag/issues/1663
For now, just use 2.1

dougbu replied:
Do you mean leave this block as-is or add `<Command Condition="'%(TargetFramework)' == 'netcoreapp2.2'">$(NSwagExe_Core21)</Command>` i.e. use 2.1 for a 2.2 web site?

Either is :ok: with me.

RSuter replied:
I think support will be better if you'd add 2.2 and use NSwagExe_Core21. Is this even working?

dougbu replied:
Done and yes it works

dougbu commented on src/NSwag.MSBuild.CodeGeneration/build.bat:
```diff
@@ -0,0 +1,3 @@
+rmdir Output /Q /S nonemptydir
```
@RSuter ??

RSuter replied:
I dont understand your question

RSuter replied:
I think this file is not needed... 01_build.bat will pack it and AppVeyor will pick the package up...

dougbu replied:
Excellent. I'll remove these two files in the next iteration.

dougbu commented:
> If the command fails, will it write all output for debugging?

Yes, because your tool writes to `stdout`, everything is passed through to the console by default.

Lesser quibble: I'm not sure if the tool's output will show up in an MSBuild log file.

RSuter commented on src/NSwag.MSBuild.CodeGeneration/NSwag.MSBuild.CodeGeneration.nuspec:
```diff
@@ -0,0 +1,24 @@
+    <tags>Swagger Documentation WebApi AspNet TypeScript CodeGen</tags>
+    <developmentDependency>true</developmentDependency>
+    <dependencies>
+      <dependency id="Microsoft.Extensions.ApiDescription.Design" Version="2.2.0-preview3-t007163513" />
```
Will a nuget.org push work or is it validation for that?

dougbu replied:
Nothing breaks when building this repo. But, users won't be able to consume the new package until the versions line up on NuGet.org. 

I'll update this to use a version range. Unfortunately, a version range here mainly helps initially because NuGet will pick the _lowest_ version it finds that's in the range. Users would need to reference `Microsoft.Extensions.ApiDescription.Design` directly in their `.csproj` file to get something newer e.g. a nightly build from https://dotnet.myget.org/gallery/aspnetcore-dev.

RSuter replied:
> I'll update this to use a version range. 

Will you update this in this PR or in a new PR?

dougbu replied:
Unfortunately, version ranges don't work with preview releases due to a NuGet bug. But, I'll update this PR to use a version available on http://dotnet.myget.org

RSuter commented:
PR ready. Wait for other PR to be ready.

RSuter commented:
@dougbu is this now ready to be merged?

dougbu commented:
:up::date: just to rebase. Still need a package version that's complete and available on NuGet.org.

RSuter commented:
Ok. PR blocked until the package is publicly available... otherwise ready to be merged.

dougbu commented:
:up:date: to resolve conflict. Still blocked for now (new package coming soon!)&hellip;

RSuter commented:
Do you have an ETA? Otherwise ill release v12 and this package later

dougbu commented:
Sorry I've been quiet @RSuter. I expect we should be publishing the package next week.

dougbu commented:
:up::date: Rebased on current `master` and updated to use NuGet.org version of Microsoft.Extensions.ApiDescription.Design package. PR is no longer blocked.

dougbu commented:
> PR is no longer blocked.

@RSuter just making sure my "bit delayed" apology didn't bury the fact this PR is read to go.

RSuter commented:
No worries, this package canbe released anytime... :-) i’m just wondering whether “NSwag.MSBuild.CodeGeneration” is the correct name as it also generates swagger documents not only code... do you have an (internal), better name for this “functionality”?

dougbu commented:
> as it also generates swagger documents not only code

Well, the end-to-end feature is about code generation. It's not currently possible to stop `<ServiceUriReference>` or `<ServiceProjectReference>` item processing after downloading a Swagger or Open API document.
> do you have an (internal), better name for this “functionality”?

Nothing significantly better. We usually refer to the feature as "client code generation" or "service reference".

dougbu commented:
> also output the swagger.json to a file

This already happens though the output directory is the client project by default. My comment was about _only_ generating the document.

RSuter commented:
Ah ok, then everything is fine :-). Please ping me if there are any issues regarding the integration of the feature... Are there any open tasks?

dougbu commented:
> Are there any open tasks?

We'll continue to improve the glue code but I don't think there's anything to address on your side.

FYI https://github.com/aspnet/Mvc/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-code-generation
